lib/Dialect/Arc/Transforms/LegalizeStateUpdate.cpp

```diff
@@ -412,6 +415,119 @@ LogicalResult Legalizer::visitBlock(Block *block) {
+  // Collect memory values and their reads
+  DenseMap<Value, SetVector<Operation *>> readOps;
+  auto result = region.walk([&](Operation *op) {
+    if (isa<MemoryWriteOp>(op))
+      return WalkResult::advance();
+    SmallVector<Value> memoriesReadFrom;
+    if (auto readOp = dyn_cast<MemoryReadOp>(op)) {
+      memoriesReadFrom.push_back(readOp.getMemory());
+    } else {
+      for (auto operand : op->getOperands())
+        if (isa<MemoryType>(operand.getType()))
+          memoriesReadFrom.push_back(operand);
+    }
+    for (auto memVal : memoriesReadFrom) {
+      if (!memories.contains(memVal))
+        return op->emitOpError("uses memory value not directly defined by a "
+                               "arc.alloc_memory operation"),
+               WalkResult::interrupt();
+      readOps[memVal].insert(op);
+    }
+
+    return WalkResult::advance();
+  });
+
+  if (result.wasInterrupted())
+    return failure();
+
+  // Collect all writes
+  SmallVector<MemoryWriteOp> writes;
+  region.walk([&](MemoryWriteOp writeOp) { writes.push_back(writeOp); });
+
+  // Move the writes
+  for (auto writeOp : writes) {
+    if (!memories.contains(writeOp.getMemory()))
+      return writeOp->emitOpError("uses memory value not directly defined by a "
+                                  "arc.alloc_memory operation");
+    for (auto *readOp : readOps[writeOp.getMemory()]) {
+      // (1) If the last read and the write are in the same block, just move the
+      // write after the read.
+      // (2) If the write is directly in the clock tree region and the last read
+      // in some nested region, move the write after the operation with the
+      // nested region. (3) If the write is nested in if-statements (arbitrarily
+      // deep) without return value, move the whole if operation after the last
+      // read or the operation that defines the region if the read is inside a
+      // nested region. (4) Number (3) may move more memory operations with the
+      // write op, thus messing up the order of previously moved memory writes,
+      // we check in a second walk-through if that is the case and just emit an
+      // error for now. We could instead move reads in a parent region, split if
+      // operations such that the memory write has its own, etc. Alternatively,
+      // rewrite this to insert temporaries which is more difficult for memories
+      // than simple states because the memory addresses have to be considered
+      // (we cannot just copy the whole memory each time).
+      Operation *readAncestor, *writeAncestor;
+      if (failed(getAncestorOpsInCommonDominatorBlock(
+              writeOp, &writeAncestor, readOp, &readAncestor, domInfo)))
+        return failure();
+      // FIXME: the 'isBeforeInBlock` + 'moveAfter' compination can be
+      // computationally very expensive.
+      if (writeAncestor->isBeforeInBlock(readAncestor))
+        writeAncestor->moveAfter(readAncestor);
+    }
```

Comment:
This looks like a reasonable approach for now! When we overhaul this pass with a custom dataflow analysis, we'll probably be able to just treat memory reads/writes exactly like state reads/writes, and move everything in one go.

Reply:
Yea, that's exactly what I'm hoping for. Didn't want to spend an unreasonable amount of time doing the memory legalization perfectly when we're going to overhaul it anyways, but also didn't want to leave the lowering pipeline boken, of course. 🙂